src/dos/program_more.cpp

```diff
@@ -0,0 +1,343 @@
+}
+
+MORE::Decision MORE::ProcessInputStream()
+{
+    bool     new_line        = false;
+	uint16_t line_counter    = 0;
+	auto     previous_column = GetCurrentColumn();
+
+	skip_cr = false;
+	skip_lf = false;
+
+    Decision decision = Decision::NextFile;
+	while (true) {
+		if (shutdown_requested) {
+			decision = Decision::Terminate;
+			break;
+		}
+
+		// Read character
+		char code = 0;
+		if (!GetCharacter(code)) {
+			decision = Decision::NextFile; // end of file
+			break;
+		}
+
+		// A trick to make it more resistant to ANSI cursor movements
+		const auto current_row = GetCurrentRow();
+		if (line_counter > current_row)
+			line_counter = current_row;
+
+		// Handle new line characters
+		if (code == code_cr) {
+			skip_lf  = true;
+			new_line = true;
+		} else if (code == code_lf) {
+			skip_cr  = true;
+			new_line = true;
+		} else {
+			skip_cr  = false;
+			skip_lf  = false;
+			new_line = false;
+		}
+
+		// Duplicate character on the output
+		if (new_line)
+			code = '\n';
+		WriteOut("%c", code);
+
+		// Detect 'new line' due to character passing the last column
+		const auto current_column = GetCurrentColumn();
+		if (!current_column && previous_column)
+			new_line = true;
+		previous_column = current_column;
+
+		// Update new line counter, decide if pause needed
+		if (new_line && current_row)
+			++line_counter;
+		if (line_counter < max_lines)
+			continue;
+
+		// New line occured just enough times for a pause
+		line_counter = 0;
+		decision = PromptUser();
+		if (decision == Decision::Terminate ||
+			decision == Decision::NextFile) {
+			new_line = false;
+			break;
+		}
+
+		// Start next screen
+		WriteOut("\n\n");
+		skip_cr = true;
+		skip_lf = true;
+	}
+
+	if (!new_line)
+		WriteOut("\n");
+
+	if (decision == Decision::Terminate) {
+		WriteOut("\n");
+		WriteOut(MSG_Get("SHELL_CMD_MORE_TERMINATED"));
+	}
+
+	return decision;
+}
+
+MORE::Decision MORE::PromptUser()
+{
+	WriteOut("\n");
+	if (mode_multifile)
+		WriteOut(MSG_Get("SHELL_CMD_MORE_PROMPT_FILE"));
+	else
+		WriteOut(MSG_Get("SHELL_CMD_MORE_PROMPT"));
+
+	while (!shutdown_requested) {
+		CALLBACK_Idle();
+
+		uint16_t count = 1;
+		char choice = 0;
+		DOS_ReadFile(STDIN, reinterpret_cast<uint8_t *>(&choice), &count);
+
+		if (count == 0)
+			return Decision::Terminate;
+
+		if (choice == code_ctrl_c)
+			WriteOut("^C");
+
+		if (choice == code_ctrl_c || choice == code_esc)
+			return Decision::Terminate;
+
+		if (choice == code_cr || choice == ' ')
+			return Decision::More;
+
+		if (!mode_multifile)
+			continue;
+
+		if (choice == 'N' || choice == 'n')
+			return Decision::NextFile;
+	}
+
+	return Decision::Terminate;
+}
+
+bool MORE::GetCharacter(char &code)
+{
+	static const std::vector<char> break_sequence = {
+		// CTRL + ZXCVBNM - emergency break key sequence, in case
+		// user executes MORE.COM with STDIN as input; supported
+		// at least by MS-DOS 6.22, DR-DOS 7.03 and PC-DOS 2000
+		0x1a, 0x18, 0x03, 0x16, 0x02, 0x0e, 0x0d
+	};
```

Comment:
Suggest:
``` c++
constexpr std::array<char, 7> break_sequence = { ... };
```

When we want to lock-in a container, `std::array` lets us use `constexpr` and gives all the functional benefits of a run-time vector (with `.begin/.end()`, `.size()`, etc..)

Reply:
OK, changed.